### ISSUE TITLE
key point visibility threshold

### ISSUE THREAD

**Issue by LjIA26** [open-mmlab/mmpose#1982]
Hello,
Is there a reason why the key points visibility threshold is set to 0.3? Instead of 0.9?.

**Comment by ly015** [comment]
The setting of these hyperparameters is usually empirical or adapted from previous works. You may need to change it to achieve the best performance in your scenario. Please note that there could be multiple keypoint visibility thresholds for different purposes like NMS and visualization.

**Comment by LjIA26** [comment]
Hello,
That's a good point,

I wasn't able to find where the threshold was for the accuracies 
precision and recall calculations during validation. Could you pinpoint me to it ?

**Comment by ly015** [comment]
In MMPose v1.0 the threshold can be found in the arguments of the evaluation metric, like in the [CocoMetric](https://github.com/open-mmlab/mmpose/blob/dev-1.x/mmpose/evaluation/metrics/coco_metric.py#L93). And in a config file it is set in the `val_evaluator` or `test_evaluator` field like [here](https://github.com/open-mmlab/mmpose/blob/dev-1.x/configs/body_2d_keypoint/topdown_heatmap/coco/td-hm_hrnet-w32_8xb64-210e_coco-256x192.py#L146-L150) (in this example the default keypoint threshold is used).

**Comment by skanelo** [comment]
Hello, first of all, thank you for the remarkable work!

I was wondering where I can find more info regarding the importance of the keypoint threshold, namely what a low threshold indicates and what I should roughly expect from a low keypoint threshold (the same goes for a higher threshold value).

Additionally, I noticed my custom model struggles for instances that are partially excluded from the image (and thus there are keypoints that cannot be annotated). In such cases, the model desperately tries to assign the missing keypoints to the image (which does not make sense since they are not part of the image) and the final result ends up having entirely wrong predictions. 
I was wondering how the model manipulates such cases and what are the best practices to tackle cases like these.

**Comment by ly015** [comment]
@skanelo Thanks for using MMPose. We provide a [visualization tool](https://mmpose.readthedocs.io/en/dev-1.x/user_guides/visualization.html) that can show the predicted heatmap to give an intuitive view of the keypoint score distribution. The score generally reflects the confidence that the keypoint is located at the predicted position, and can be affected by many factors like occlusion, the scale of the target person, rare poses, image resolution or quality, et al. So it's often difficult to determine whether a keypoint is occluded by setting a score threshold.

One possible way to alleviate the occlusion problem is to annotate those invisible keypoints empirically in the training data (like the COCO dataset), so the model can give more reliable predictions in occluded cases. Another idea is to use an extra head to predict keypoint visibility as a binary classification task, which should be more accurate than thresholding the keypoint score.


![image](https://user-images.githubusercontent.com/15977946/220333810-e86f05b9-b7f4-488f-8aca-879ed4a2d798.png)

**Comment by skanelo** [comment]
@ly015 Thank you very much for your prompt and insightful response. 

Just one clarification, I was referring to keypoints that are **out of the image**, as a result, they cannot be annotated at all. I have created a custom script to visualize the predictions of the model so I was thinking to visualize the confidence score for each keypoint and examine these specific cases (in which the object of interest lies partially out of the image). I'd expect the mistakenly assigned keypoints to have a lower confidence threshold, and if that's the case, I guess by lowering it I could alleviate this issue. What do you think? 

What you suggested works really well indeed (cases in which the keypoints are occluded by another object have been annotated empirically and it **does** help the model to detect them)!

**Comment by ly015** [comment]
Yes, if the keypoint is outside the image, the corresponding score should be significantly lower than in normal cases and can be filtered by a threshold. Of course, the converse is not necessarily true, that is, a predicted keypoint with a low score could be caused by many reasons.

And in some hard crowd or occlusion cases, it can be difficult to determine whether a keypoint is outside the image by score. For example, if there are 2 overlapped people in the image, one's hand is outside the image but the other's hand is visible, the model may not be able to assign the keypoint to the person correctly in this case.

**Comment by skanelo** [comment]
I see, thank you very much for your detailed response!!!
